add specialized InList implementations for common scalar types

adriangb opened the pull request:
Closes #18824

adriangb commented:
Here's what I'm seeing so far:
Function & State | main (us) | specialized (us) | Change
-- | -- | -- | --
in_list_utf8(5) (1024, 0) IN (1, 0) | 3.6205 | 4.8594 | Regressed (+34.239%)
in_list_utf8(10) (1024, 0) IN (1, 0) | 3.6363 | 5.1001 | Regressed (+40.312%)
in_list_utf8(20) (1024, 0) IN (1, 0) | 3.6707 | 4.9045 | Regressed (+33.600%)
in_list_f32 (1024, 0) IN (1, 0) | 3.5255 | 3.2288 | Improved (−8.2971%)
in_list_i32 (1024, 0) IN (1, 0) | 3.4923 | 3.6403 | Regressed (+3.8772%)
in_list_utf8(5) (1024, 0.2) IN (1, 0) | 4.9284 | 6.3583 | Regressed (+29.002%)
in_list_utf8(10) (1024, 0.2) IN (1, 0) | 4.9788 | 6.3273 | Regressed (+27.064%)
in_list_utf8(20) (1024, 0.2) IN (1, 0) | 5.1145 | 6.4905 | Regressed (+26.907%)
in_list_f32 (1024, 0.2) IN (1, 0) | 4.4894 | 4.2540 | Improved (−5.2512%)
in_list_i32 (1024, 0.2) IN (1, 0) | 4.4055 | 4.2869 | Improved (−2.7526%)
in_list_utf8(5) (1024, 0) IN (3, 0) | 3.5576 | 4.8733 | Regressed (+37.000%)
in_list_utf8(10) (1024, 0) IN (3, 0) | 3.4999 | 4.8737 | Regressed (+37.108%)
in_list_utf8(20) (1024, 0) IN (3, 0) | 3.4897 | 4.9039 | Regressed (+40.449%)
in_list_f32 (1024, 0) IN (3, 0) | 3.5114 | 3.2550 | Improved (−7.3270%)
in_list_i32 (1024, 0) IN (3, 0) | 3.5199 | 3.4879 | Noise Threshold (−0.9158%)
in_list_utf8(5) (1024, 0.2) IN (3, 0) | 5.0196 | 6.2232 | Regressed (+23.865%)
in_list_utf8(10) (1024, 0.2) IN (3, 0) | 5.0453 | 6.5280 | Regressed (+28.004%)
in_list_utf8(20) (1024, 0.2) IN (3, 0) | 5.1005 | 11.124 | Regressed (+118.13%)
in_list_f32 (1024, 0.2) IN (3, 0) | 4.5049 | 4.3126 | Improved (−4.6367%)
in_list_i32 (1024, 0.2) IN (3, 0) | 4.4026 | 4.4424 | Noise Threshold (+1.0449%)
in_list_utf8(5) (1024, 0) IN (10, 0) | 3.5275 | 4.9709 | Regressed (+40.969%)
in_list_utf8(10) (1024, 0) IN (10, 0) | 3.5613 | 4.9125 | Regressed (+37.386%)
in_list_utf8(20) (1024, 0) IN (10, 0) | 3.5589 | 4.8611 | Regressed (+36.578%)
in_list_f32 (1024, 0) IN (10, 0) | 3.5596 | 3.2017 | Improved (−10.003%)
in_list_i32 (1024, 0) IN (10, 0) | 3.4917 | 3.5190 | Noise Threshold (+1.0216%)
in_list_utf8(5) (1024, 0.2) IN (10, 0) | 5.0368 | 6.6162 | Regressed (+31.725%)
in_list_utf8(10) (1024, 0.2) IN (10, 0) | 5.0980 | 6.6660 | Regressed (+30.039%)
in_list_utf8(20) (1024, 0.2) IN (10, 0) | 5.2543 | 6.5350 | Regressed (+24.335%)
in_list_f32 (1024, 0.2) IN (10, 0) | 4.5609 | 4.3248 | Improved (−5.3746%)
in_list_i32 (1024, 0.2) IN (10, 0) | 4.4460 | 4.3543 | Improved (−2.7754%)
in_list_utf8(5) (1024, 0) IN (100, 0) | 3.6542 | 4.9992 | Regressed (+36.952%)
in_list_utf8(10) (1024, 0) IN (100, 0) | 3.6529 | 4.8772 | Regressed (+33.560%)
in_list_utf8(20) (1024, 0) IN (100, 0) | 3.6155 | 5.0250 | Regressed (+39.127%)
in_list_f32 (1024, 0) IN (100, 0) | 3.6029 | 3.2448 | Improved (−9.7603%)
in_list_i32 (1024, 0) IN (100, 0) | 3.6048 | 3.4770 | Improved (−3.5307%)
in_list_utf8(5) (1024, 0.2) IN (100, 0) | 5.3988 | 6.6108 | Regressed (+22.866%)
in_list_utf8(10) (1024, 0.2) IN (100, 0) | 5.4776 | 6.6567 | Regressed (+21.591%)
in_list_utf8(20) (1024, 0.2) IN (100, 0) | 5.3721 | 6.6470 | Regressed (+24.058%)
in_list_f32 (1024, 0.2) IN (100, 0) | 4.7693 | 4.3281 | Improved (−9.9814%)
in_list_i32 (1024, 0.2) IN (100, 0) | 4.5402 | 4.4076 | Improved (−3.5247%)

I think we'd need to add benchmarks for other primitive types.
And it's interesting that Utf8 regresses a lot across the board. I guess the benefits of vectorization / computing the entire hashes at once outweighs the dynamic dispatch?

martin-g commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -198,18 +208,135 @@ impl ArrayStaticFilter {
         let in_array = in_array
-            .as_primitive_opt::<Int32Type>()
+            .as_boolean_opt()
             .ok_or_else(|| exec_datafusion_err!("Failed to downcast array"))?;
```
```suggestion
            .ok_or_else(|| exec_datafusion_err!("Failed to downcast an array to a boolean array"))?;
```

adriangb commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -198,68 +206,122 @@ impl ArrayStaticFilter {
+                    }
+                    (false, false, false) => {
+                        // no nulls anywhere, not negated
+                        BooleanArray::from_iter(
```
I do wonder why we don't have faster high-level APIs if this is really important. E.g. `BooleanArray::new_false`, `BooleanArray::new_nulls`, `BooleanArray::new_true` and `BooleanArray::collect_bool(size, iterator)` or something like that.

alamb commented:
🤖: Benchmark completed

<details><summary>Details</summary>
<p>


```
group                                       main                                   specialize
-----                                       ----                                   ----------
in_list_f32 (1024, 0) IN (1, 0)             1.16      4.2±0.01µs        ? ?/sec    1.00      3.6±0.02µs        ? ?/sec
in_list_f32 (1024, 0) IN (10, 0)            1.16      4.2±0.01µs        ? ?/sec    1.00      3.6±0.02µs        ? ?/sec
in_list_f32 (1024, 0) IN (100, 0)           1.16      4.3±0.01µs        ? ?/sec    1.00      3.7±0.01µs        ? ?/sec
in_list_f32 (1024, 0) IN (3, 0)             1.17      4.3±0.10µs        ? ?/sec    1.00      3.6±0.01µs        ? ?/sec
in_list_f32 (1024, 0.2) IN (1, 0)           1.00      5.1±0.02µs        ? ?/sec    1.13      5.8±0.07µs        ? ?/sec
in_list_f32 (1024, 0.2) IN (10, 0)          1.00      5.3±0.03µs        ? ?/sec    1.09      5.7±0.02µs        ? ?/sec
in_list_f32 (1024, 0.2) IN (100, 0)         1.00      5.3±0.04µs        ? ?/sec    1.13      5.9±0.10µs        ? ?/sec
in_list_f32 (1024, 0.2) IN (3, 0)           1.00      5.2±0.04µs        ? ?/sec    1.12      5.8±0.05µs        ? ?/sec
in_list_i32 (1024, 0) IN (1, 0)             1.00      4.3±0.01µs        ? ?/sec    1.11      4.7±0.02µs        ? ?/sec
in_list_i32 (1024, 0) IN (10, 0)            1.00      4.2±0.01µs        ? ?/sec    1.11      4.7±0.04µs        ? ?/sec
in_list_i32 (1024, 0) IN (100, 0)           1.00      4.3±0.01µs        ? ?/sec    1.11      4.7±0.01µs        ? ?/sec
in_list_i32 (1024, 0) IN (3, 0)             1.00      4.3±0.01µs        ? ?/sec    1.11      4.7±0.07µs        ? ?/sec
in_list_i32 (1024, 0.2) IN (1, 0)           1.00      5.2±0.02µs        ? ?/sec    1.32      6.8±0.39µs        ? ?/sec
in_list_i32 (1024, 0.2) IN (10, 0)          1.00      5.2±0.02µs        ? ?/sec    1.25      6.5±0.03µs        ? ?/sec
in_list_i32 (1024, 0.2) IN (100, 0)         1.00      5.2±0.03µs        ? ?/sec    1.16      6.1±0.06µs        ? ?/sec
in_list_i32 (1024, 0.2) IN (3, 0)           1.00      5.2±0.02µs        ? ?/sec    1.30      6.8±0.03µs        ? ?/sec
in_list_utf8(10) (1024, 0) IN (1, 0)        1.16      4.5±0.02µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(10) (1024, 0) IN (10, 0)       1.16      4.5±0.01µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(10) (1024, 0) IN (100, 0)      1.16      4.5±0.03µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(10) (1024, 0) IN (3, 0)        1.16      4.5±0.04µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(10) (1024, 0.2) IN (1, 0)      1.06      5.9±0.03µs        ? ?/sec    1.00      5.6±0.04µs        ? ?/sec
in_list_utf8(10) (1024, 0.2) IN (10, 0)     1.09      6.0±0.06µs        ? ?/sec    1.00      5.5±0.03µs        ? ?/sec
in_list_utf8(10) (1024, 0.2) IN (100, 0)    1.07      6.0±0.03µs        ? ?/sec    1.00      5.6±0.03µs        ? ?/sec
in_list_utf8(10) (1024, 0.2) IN (3, 0)      1.10      6.1±0.04µs        ? ?/sec    1.00      5.6±0.03µs        ? ?/sec
in_list_utf8(20) (1024, 0) IN (1, 0)        1.16      4.5±0.02µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(20) (1024, 0) IN (10, 0)       1.16      4.5±0.01µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(20) (1024, 0) IN (100, 0)      1.15      4.5±0.04µs        ? ?/sec    1.00      3.9±0.03µs        ? ?/sec
in_list_utf8(20) (1024, 0) IN (3, 0)        1.14      4.6±0.01µs        ? ?/sec    1.00      4.0±0.03µs        ? ?/sec
in_list_utf8(20) (1024, 0.2) IN (1, 0)      1.06      6.0±0.02µs        ? ?/sec    1.00      5.7±0.03µs        ? ?/sec
in_list_utf8(20) (1024, 0.2) IN (10, 0)     1.07      6.0±0.03µs        ? ?/sec    1.00      5.6±0.03µs        ? ?/sec
in_list_utf8(20) (1024, 0.2) IN (100, 0)    1.09      6.1±0.03µs        ? ?/sec    1.00      5.6±0.02µs        ? ?/sec
in_list_utf8(20) (1024, 0.2) IN (3, 0)      1.04      6.1±0.03µs        ? ?/sec    1.00      5.8±0.05µs        ? ?/sec
in_list_utf8(5) (1024, 0) IN (1, 0)         1.15      4.5±0.01µs        ? ?/sec    1.00      4.0±0.01µs        ? ?/sec
in_list_utf8(5) (1024, 0) IN (10, 0)        1.16      4.5±0.03µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
in_list_utf8(5) (1024, 0) IN (100, 0)       1.16      4.5±0.01µs        ? ?/sec    1.00      3.9±0.00µs        ? ?/sec
in_list_utf8(5) (1024, 0) IN (3, 0)         1.14      4.5±0.03µs        ? ?/sec    1.00      4.0±0.02µs        ? ?/sec
in_list_utf8(5) (1024, 0.2) IN (1, 0)       1.06      5.9±0.04µs        ? ?/sec    1.00      5.5±0.04µs        ? ?/sec
in_list_utf8(5) (1024, 0.2) IN (10, 0)      1.10      6.1±0.04µs        ? ?/sec    1.00      5.5±0.03µs        ? ?/sec
in_list_utf8(5) (1024, 0.2) IN (100, 0)     1.11      6.0±0.03µs        ? ?/sec    1.00      5.4±0.03µs        ? ?/sec
in_list_utf8(5) (1024, 0.2) IN (3, 0)       1.11      6.1±0.06µs        ? ?/sec    1.00      5.5±0.04µs        ? ?/sec
```


</p>
</details>

Dandandan commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -198,68 +211,127 @@ impl ArrayStaticFilter {
+                    }
+                    (true, _, true) | (false, true, true) => {
+                        // Either needle or haystack has nulls, negated
+                        BooleanArray::from_iter(v.iter().map(|value| {
```
It probably would be faster to handle the nulls separately or using `set_indices` rather than using `BooleanArray::from_iter` and `v.iter` etc.

Dandandan commented:
> 🤖: Benchmark completed
> 
> Details
> 
> ```
> group                                       main                                   specialize
> -----                                       ----                                   ----------
> in_list_f32 (1024, 0) IN (1, 0)             1.16      4.2±0.01µs        ? ?/sec    1.00      3.6±0.02µs        ? ?/sec
> in_list_f32 (1024, 0) IN (10, 0)            1.16      4.2±0.01µs        ? ?/sec    1.00      3.6±0.02µs        ? ?/sec
> in_list_f32 (1024, 0) IN (100, 0)           1.16      4.3±0.01µs        ? ?/sec    1.00      3.7±0.01µs        ? ?/sec
> in_list_f32 (1024, 0) IN (3, 0)             1.17      4.3±0.10µs        ? ?/sec    1.00      3.6±0.01µs        ? ?/sec
> in_list_f32 (1024, 0.2) IN (1, 0)           1.00      5.1±0.02µs        ? ?/sec    1.13      5.8±0.07µs        ? ?/sec
> in_list_f32 (1024, 0.2) IN (10, 0)          1.00      5.3±0.03µs        ? ?/sec    1.09      5.7±0.02µs        ? ?/sec
> in_list_f32 (1024, 0.2) IN (100, 0)         1.00      5.3±0.04µs        ? ?/sec    1.13      5.9±0.10µs        ? ?/sec
> in_list_f32 (1024, 0.2) IN (3, 0)           1.00      5.2±0.04µs        ? ?/sec    1.12      5.8±0.05µs        ? ?/sec
> in_list_i32 (1024, 0) IN (1, 0)             1.00      4.3±0.01µs        ? ?/sec    1.11      4.7±0.02µs        ? ?/sec
> in_list_i32 (1024, 0) IN (10, 0)            1.00      4.2±0.01µs        ? ?/sec    1.11      4.7±0.04µs        ? ?/sec
> in_list_i32 (1024, 0) IN (100, 0)           1.00      4.3±0.01µs        ? ?/sec    1.11      4.7±0.01µs        ? ?/sec
> in_list_i32 (1024, 0) IN (3, 0)             1.00      4.3±0.01µs        ? ?/sec    1.11      4.7±0.07µs        ? ?/sec
> in_list_i32 (1024, 0.2) IN (1, 0)           1.00      5.2±0.02µs        ? ?/sec    1.32      6.8±0.39µs        ? ?/sec
> in_list_i32 (1024, 0.2) IN (10, 0)          1.00      5.2±0.02µs        ? ?/sec    1.25      6.5±0.03µs        ? ?/sec
> in_list_i32 (1024, 0.2) IN (100, 0)         1.00      5.2±0.03µs        ? ?/sec    1.16      6.1±0.06µs        ? ?/sec
> in_list_i32 (1024, 0.2) IN (3, 0)           1.00      5.2±0.02µs        ? ?/sec    1.30      6.8±0.03µs        ? ?/sec
> in_list_utf8(10) (1024, 0) IN (1, 0)        1.16      4.5±0.02µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(10) (1024, 0) IN (10, 0)       1.16      4.5±0.01µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(10) (1024, 0) IN (100, 0)      1.16      4.5±0.03µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(10) (1024, 0) IN (3, 0)        1.16      4.5±0.04µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(10) (1024, 0.2) IN (1, 0)      1.06      5.9±0.03µs        ? ?/sec    1.00      5.6±0.04µs        ? ?/sec
> in_list_utf8(10) (1024, 0.2) IN (10, 0)     1.09      6.0±0.06µs        ? ?/sec    1.00      5.5±0.03µs        ? ?/sec
> in_list_utf8(10) (1024, 0.2) IN (100, 0)    1.07      6.0±0.03µs        ? ?/sec    1.00      5.6±0.03µs        ? ?/sec
> in_list_utf8(10) (1024, 0.2) IN (3, 0)      1.10      6.1±0.04µs        ? ?/sec    1.00      5.6±0.03µs        ? ?/sec
> in_list_utf8(20) (1024, 0) IN (1, 0)        1.16      4.5±0.02µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(20) (1024, 0) IN (10, 0)       1.16      4.5±0.01µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(20) (1024, 0) IN (100, 0)      1.15      4.5±0.04µs        ? ?/sec    1.00      3.9±0.03µs        ? ?/sec
> in_list_utf8(20) (1024, 0) IN (3, 0)        1.14      4.6±0.01µs        ? ?/sec    1.00      4.0±0.03µs        ? ?/sec
> in_list_utf8(20) (1024, 0.2) IN (1, 0)      1.06      6.0±0.02µs        ? ?/sec    1.00      5.7±0.03µs        ? ?/sec
> in_list_utf8(20) (1024, 0.2) IN (10, 0)     1.07      6.0±0.03µs        ? ?/sec    1.00      5.6±0.03µs        ? ?/sec
> in_list_utf8(20) (1024, 0.2) IN (100, 0)    1.09      6.1±0.03µs        ? ?/sec    1.00      5.6±0.02µs        ? ?/sec
> in_list_utf8(20) (1024, 0.2) IN (3, 0)      1.04      6.1±0.03µs        ? ?/sec    1.00      5.8±0.05µs        ? ?/sec
> in_list_utf8(5) (1024, 0) IN (1, 0)         1.15      4.5±0.01µs        ? ?/sec    1.00      4.0±0.01µs        ? ?/sec
> in_list_utf8(5) (1024, 0) IN (10, 0)        1.16      4.5±0.03µs        ? ?/sec    1.00      3.9±0.01µs        ? ?/sec
> in_list_utf8(5) (1024, 0) IN (100, 0)       1.16      4.5±0.01µs        ? ?/sec    1.00      3.9±0.00µs        ? ?/sec
> in_list_utf8(5) (1024, 0) IN (3, 0)         1.14      4.5±0.03µs        ? ?/sec    1.00      4.0±0.02µs        ? ?/sec
> in_list_utf8(5) (1024, 0.2) IN (1, 0)       1.06      5.9±0.04µs        ? ?/sec    1.00      5.5±0.04µs        ? ?/sec
> in_list_utf8(5) (1024, 0.2) IN (10, 0)      1.10      6.1±0.04µs        ? ?/sec    1.00      5.5±0.03µs        ? ?/sec
> in_list_utf8(5) (1024, 0.2) IN (100, 0)     1.11      6.0±0.03µs        ? ?/sec    1.00      5.4±0.03µs        ? ?/sec
> in_list_utf8(5) (1024, 0.2) IN (3, 0)       1.11      6.1±0.06µs        ? ?/sec    1.00      5.5±0.04µs        ? ?/sec
> ```

Results seem a bit mixed?

adriangb commented:
I realized this new implementation differs from the existing one in ways which fix some bugs. I opened https://github.com/apache/datafusion/pull/19050 to demonstrate those bugs. So maybe lets first fix the bugs and merge the new tests, then we can continue here?

geoffreyclaude commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -198,65 +248,414 @@ impl ArrayStaticFilter {
-        }
+    #[inline]
+    fn contains(&self, value: &T) -> bool {
+        self.0.binary_search(value).is_ok()
```
I wanted to do a bench of linear vs binary vs hash to see if thresholds come up

adriangb replied:
> AI coding agents are really super useful for quickly bringing crazy ideas to life, I'm a big fan of our current timeline 😄

Poor @alamb 😭

geoffreyclaude replied:
> smaller targeted PRs off of that

That's definitely the way to go! I'll fork off an issue from https://github.com/apache/datafusion/issues/18824.

adriangb commented:
run benchmarks in_list

adriangb commented:
run benchmark in_list

alamb commented:
> ~ hey real @alamb could you whitelist `in_list` for me (might as well do Daniel and Geoffrey at the same time) ~ nvm user error

@Dandandan  is already on the list
https://github.com/alamb/datafusion-benchmarking/blob/4fb120785fa66ecbf40a45d8a5d0d5f4be17266a/scripts/scrape_comments.py#L41

I can add @geoffreyclaude if he would like

adriangb commented:
The key is `benchmark` vs `benchmarks` fyi. I can make a PR to accept both tomorrow.

adriangb commented:
run benchmark in_list

adriangb commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -198,99 +213,302 @@ impl ArrayStaticFilter {
+                        needle_nulls.cloned()
+                    }
+                    (false, true) => {
+                        // Only haystack has nulls - null where not-in-set
```
0b5a9c9

geoffreyclaude replied:
Thanks, this is way more readable!

adriangb commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -198,99 +213,302 @@ impl ArrayStaticFilter {
-            .ok_or_else(|| exec_datafusion_err!("Failed to downcast array"))?;
+impl PartialEq for OrderedFloat32 {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.total_cmp(&other.0).is_eq()
```
b6404e5bc

adriangb commented:
run benchmark in_list

adriangb commented:
The benchmarks are now looking *very* good, and this has gone through several rounds of review, all of it addressed. But there are significant changes since the last approvals. @geoffreyclaude could you give another review and approve if you think it's ready so we can merge this piece and continue work in https://github.com/apache/datafusion/issues/19241

geoffreyclaude commented:
> The benchmarks are now looking *very* good, and this has gone through several rounds of review, all of it addressed. But there are significant changes since the last approvals. @geoffreyclaude could you give another review and approve if you think it's ready so we can merge this piece and continue work in https://github.com/apache/datafusion/issues/19241

The tests you added as a dedicated PR give a lot of confidence this isn't introducing any functional regression. And perf-wise, then benchmarks speak for themselves!

Do we have more generic benches that exercise this path? Maybe the Clickbench ones? Would be nice to see the big picture and prove this isn't "benchmaxing" with unexpected adverse effects in real life.

Otherwise, ✅ from me of course. You addressed my remaining two nits already.

adriangb commented:
I'll kick off some general benchmarks and if those look good and there's no more feedback I'll merge later today or tomorrow morning.

adriangb commented:
run benchmark clickbench_partitioned tpch

alamb commented on datafusion/physical-expr/src/expressions/in_list.rs:
```diff
@@ -138,9 +142,20 @@ fn instantiate_static_filter(
+        DataType::UInt64 => Ok(Arc::new(UInt64StaticFilter::try_new(&in_array)?)),
+        // Float primitive types (use ordered wrappers for Hash/Eq)
+        DataType::Float32 => Ok(Arc::new(Float32StaticFilter::try_new(&in_array)?)),
+        DataType::Float64 => Ok(Arc::new(Float64StaticFilter::try_new(&in_array)?)),
```
Do we know if adding an optimized version of the binary / string type comparisons is tracked with a ticket?
